internal/controller/phases.go

```diff
@@ -143,6 +138,83 @@ func (p *phaseReconciler) initializePhaseReconciler(ctx context.Context) (reconc
 	return reconcile.Result{}, nil
 }
 
+func (p *phaseReconciler) ensureProviderVersion(ctx context.Context) (reconcile.Result, error) {
+	spec := p.provider.GetSpec()
+
+	// Return immediately if the version is already set.
+	if spec.Version != "" {
+		return reconcile.Result{}, nil
+	}
+
+	log := ctrl.LoggerFrom(ctx)
+
+	log.Info("Setting the latest provider version.")
+
+	var repo repository.Repository
+
+	var err error
+
+	// Create repository either from provided URL or from a config map.
+	if p.provider.GetSpec().FetchConfig != nil && p.provider.GetSpec().FetchConfig.Selector != nil {
```

Comment:
can we optimize it and initialize the repo only once? It gets created in the later phases too

Reply:
yeah, I slightly rewrote the original PR, so it is more optimized now.